### ISSUE TITLE
Add a test data generator

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#8142]
The jOOQ runtime has all the meta information to describe the schema. This meta information could be leveraged to generate test data on a possibly empty database. Some interesting features:

- Users could provide explicit setup data in different formats, including the jOOQ text table representation, JSON, etc.
- Explicitly provided data will be inserted as provided
- Nullable columns without explicit setup data would be null
- Non-nullable columns without explicit setup data would have random values (check constraints would need to be respected)
- Foreign keys that point to parent tables without explicit setup data would have a parent record generated (hierarchies and cycles can be managed)

This would allow for setting up a test database with sample data just for one individual test very easily.

----
See also:
- https://groups.google.com/forum/#!topic/jooq-user/lo0IoW6mORg

**Comment by dantonini** [comment]
Lukas,
for the 5th point, it seems the most reasonable behavior but might not be the desiderated behaviour. 
According my esperience it is important define what to aim.

A scenario is having some facility to setup you data, a completely different scenario is a full dataset generation.

How many parent are generated? How many parent of parent are generated? How much time is spent to insert these data into database?
In database schema with many not null foreign keys the framework will create many objects and these data are a 99% not relevant to the test, so time wasted in generation/insert and in debugging when something goes wrong with the test.

More simpler is to use a broken foreign key value and disable foreign key when insert data. Usually my philosophy with dataset setup is: if you do not declare in setup, you don't care, so the framework do the minimum work to allow insertion (also in dirty way).

Of course if you want create a "real navigable full dataset generation" you have to handle parent creation, and ancestors creation as well, but what about cardinality of autogenerated parent objects? two records could have: different parents or same parent for example.
Same consideration for lookup tables. Usually these table have a small number of objects, but many references to it. Does the dataset generation respect this behaviour?
More complexity means more reality for random generated data, but although it could be usefull in some cases, this could be hard and time consuming to implement.

**Comment by lukaseder** [comment]
> More simpler is to use a broken foreign key value and disable foreign key when insert data. Usually my philosophy with dataset setup is: if you do not declare in setup, you don't care, so the framework do the minimum work to allow insertion (also in dirty way).

Yes, that's how I understood it. If unspecified, there will be only exactly one parent entity that all child entities point to in order to maintain referential integrity.

> Same consideration for lookup tables. Usually these table have a small number of objects, but many references to it. Does the dataset generation respect this behaviour?

I would expect there needs to be a different "area" for the setup of master data, which is setup once per test suite (and never deleted / modified), not once per test. Of course, it will be the user's responsibility to specify what's master data and what's user data.

**Comment by dantonini** [comment]
If foreign key has not null & unique constraint you have to create several objects.

Look up tables & Master/user data are quite interesting concept but we need a better clarification of scenarios. Are you interested in generate "real navigable full dataset generation", or just fill tables with data?

**Comment by yamass** [comment]
We have implemented this kind of test data generator independent of JOOQ some years ago: https://btc-ag.github.io/redg/documentation/specifying_test_data/

Would be very nice to see something like this baked into JOOQ. I personally nowadays use it in conjunction with JOOQ, but it also works for any other framework.

**Comment by lukaseder** [comment]
Thanks for your feedback, @yamass. Yes, it would be great to have this in jOOQ. It competes with a lot of other bigger tasks for a place on the roadmap's higher priority section, though :)

**Comment by alf** [comment]
I believe this would be very useful.
